lib/pbench/cli/server/shell.py

```diff
@@ -47,30 +103,59 @@ def main():
         print(e)
         sys.exit(1)
     logger = get_pbench_logger(PROG, server_config)
-    find_the_unicorn(logger)
+    if site.ENABLE_USER_SITE:
+        find_the_unicorn(logger)
     try:
         host = str(server_config.get("pbench-server", "bind_host"))
         port = str(server_config.get("pbench-server", "bind_port"))
-        db = str(server_config.get("Postgres", "db_uri"))
+        db_uri = str(server_config.get("Postgres", "db_uri"))
         workers = str(server_config.get("pbench-server", "workers"))
         worker_timeout = str(server_config.get("pbench-server", "worker_timeout"))
-        pbench_top_dir = server_config.get("pbench-server", "pbench-top-dir")
-        pbench_install = server_config.get("pbench-server", "install-dir")
-        logger.info("Pbench server using database {}", db)
-
-        # Multiple gunicorn workers will attempt to connect to the DB; rather
-        # than attempt to synchronize them, detect a missing DB (from the
-        # postgres URI) and create it here. It's safer to do this here,
-        # where we're single-threaded.
-        if not database_exists(db):
-            logger.info("Postgres DB {} doesn't exist", db)
-            create_database(db)
-            logger.info("Created DB {}", db)
-        Database.init_db(server_config, logger)
+        crontab_dir = server_config.get("pbench-server", "crontab-dir")
     except (NoOptionError, NoSectionError):
         logger.exception("Error fetching required configuration")
         sys.exit(1)
 
+    logger.info("Pbench server using database {}", db_uri)
+
+    logger.debug("Waiting for PostgreSQL to become available.")
+    try:
+        wait_for_postgresql(db_uri, 120)
+    except ConnectionRefusedError:
+        logger.error("PostgreSQL not responding")
+        sys.exit(1)
+
+    # Multiple gunicorn workers will attempt to connect to the DB; rather than
+    # attempt to synchronize them, detect a missing DB (from the postgres URI)
+    # and create it here. It's safer to do this here, where we're
+    # single-threaded.
+    if not database_exists(db_uri):
+        logger.info("Postgres DB {} doesn't exist", db_uri)
+        create_database(db_uri)
+        logger.info("Created DB {}", db_uri)
+    Database.init_db(server_config, logger)
+
+    # Create the crontab file from the server configuration.
+    os.environ["PATH"] = add_to_path(server_config.BINDIR, os.environ["PATH"])
+    cp = subprocess.run(
+        ["pbench-create-crontab", crontab_dir],
+        cwd=server_config.log_dir,
+    )
+    if cp.returncode != 0:
+        logger.error(
+            "Failed to create crontab file from configuration: {}", cp.returncode
+        )
+        sys.exit(1)
+
+    # Install the created crontab file.
+    cp = subprocess.run(
+        ["crontab", f"{crontab_dir}/crontab"], cwd=server_config.log_dir
+    )
+    if cp.returncode != 0:
+        logger.error("Failed to install crontab file: {}", cp.returncode)
+        sys.exit(1)
```

Comment:
Intriguing. I do like the idea of essentially deferring the `crontab` startup until after we've started the server and initialized the DB; I'd been worrying a bit about that in the back of my mind.

Running `pbench-create-crontab` on each server startup seems excessive; should this be protected by a check for `f"{crontab_dir}/crontab"` existance?

Also, I've never tried to simply replace a running `crontab` by just installing a new one instead of deleting the old one first. Does that work reliably? (I don't see any explicit prohibition, but ...)

Reply:
Oh, wait, that should not be in this PR.  Sorry.  I'll fix.